lib/abbreviato/truncated_sax_document.rb

```diff
@@ -0,0 +1,194 @@
+require 'nokogiri'
+require 'htmlentities'
+
+class TruncatedSaxDocument < Nokogiri::XML::SAX::Document
+  IGNORABLE_TAGS = %w[html head body].freeze
+
+  # These don't have to be closed (which also impacts ongoing length calculations)
+  # http://www.456bereastreet.com/archive/201005/void_empty_elements_and_self-closing_start_tags_in_html/
+  VOID_TAGS = %w[area base br col command hr img input keygen link meta param source wbr].freeze
+
+  attr_reader :truncated_string,
+    :max_length,
+    :tail,
+    :ignored_levels,
+    :truncated
+
+  def initialize(options)
+    @html_coder = HTMLEntities.new
+
+    @max_length = options[:max_length]
+    @tail = options[:tail] || ''
+
+    @truncated_string = ""
+    @closing_tags = []
+    @estimated_length = 0
+    @ignored_levels = 0
+    @truncated = false
+  end
+
+  # This method is called when the parser encounters an open tag
+  def start_element(name, attributes)
+    return if max_length_reached || ignorable_tag?(name)
```

Comment:
same question here about whether `@truncated` needs to be true if `max_length_reached` is true